WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptViewController.swift

```diff
@@ -0,0 +1,72 @@
+import WebKit
+
+/// Previews a backend-generated receipt
+final class ReceiptViewController: UIViewController {
+    @IBOutlet private weak var webView: WKWebView!
+    private let viewModel: ReceiptViewModel
+
+    init(viewModel: ReceiptViewModel) {
+        self.viewModel = viewModel
+        super.init(nibName: "LegacyReceiptViewController", bundle: nil)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        configureContent()
+        configureNavigation()
+    }
+
+    private func configureContent() {
+        guard let receipt = viewModel.receiptRequest else {
+            DDLogError("No receipt could be found.")
+            navigationController?.popViewController(animated: true)
+            return
+        }
+        webView.load(receipt)
+    }
+
+    private func configureNavigation() {
+        let shareButton = UIBarButtonItem(image: UIImage(systemName: "square.and.arrow.up"),
+                                          style: .plain,
+                                          target: self,
+                                          action: #selector(shareReceipt))
+        let printButton = UIBarButtonItem(image: UIImage(systemName: "printer"),
+                                          style: .plain,
+                                          target: self,
+                                          action: #selector(printReceipt))
+        navigationItem.rightBarButtonItems = [shareButton, printButton]
+    }
+
+    @objc private func shareReceipt() {
+        guard let url = URL(string: viewModel.receiptURLString) else {
+            return
+        }
+        let activityViewController = UIActivityViewController(activityItems: [url],
+                                                applicationActivities: nil)
+        present(activityViewController, animated: true)
+    }
+
+    @objc private func printReceipt() {
+        guard let url = URL(string: viewModel.receiptURLString) else {
+            return
+        }
+        let printController = UIPrintInteractionController.shared
+        let printInfo = UIPrintInfo(dictionary: nil)
+        printInfo.jobName = "Receipt"
+        printController.printInfo = printInfo
+        printController.printFormatter = webView.viewPrintFormatter()
```

Comment:
We may get more control of the paper size by becoming the delegate here and implementing `printInteractionController(_ printInteractionController: UIPrintInteractionController, choosePaper paperList: [UIPrintPaper]) -> UIPrintPaper`

Reply:
I've tried it with the following:

```
func printInteractionController(_ printInteractionController: UIPrintInteractionController, choosePaper paperList: [UIPrintPaper]) -> UIPrintPaper {
        let size = webView.viewPrintFormatter().printPageRenderer?.paperRect.size ?? .zero
        let paper = UIPrintPaper.bestPaper(forPageSize: size,
                                      withPapersFrom: paperList)
        return paper
    }
```

Which leaves the visible setting on "Auto select", but internally uses `_name: "roll_anon_102.0x0mm"`. Unfortunately, it splits the receipt in to two pages, with a gap between "Subtotal" and "Discount".

I think that's because size is hitting the nil-coalescing case, and so it's using `.zero`, and getting some minimum paper spool length or something.

If we can more accurately get a CGSize for the whole document here, it should print correctly without waste.

Reply:
> The default is the app's name, so it might be good to include that too. Given we can print shipping labels from the app, I think it would be best as "Woo: receipt for order https://github.com/woocommerce/woocommerce-ios/issues/10289 on OsomStore", with a similar alternative for shipping labels and customs forms. These should be localized too.

Agree! Fixed here: https://github.com/woocommerce/woocommerce-ios/pull/11858/commits/8d5cd342201a87611e9dd22af5cb2b14c6331220 and here: https://github.com/woocommerce/woocommerce-ios/pull/11858/commits/c25910b485dc011a5ee18913c92074a67466c111, since a store might have a name, an empty name, or be nil I've also added tests for each case, so we're sure to print the right text. Good thing is this part I can test with the simulated printer:

<img width="608" alt="Screenshot 2024-02-01 at 16 48 55" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/a541d1ec-17a3-4669-a111-bb3136cbbe76">

> We may get more control of the paper size by becoming the delegate here and implementing printInteractionController(_ printInteractionController: UIPrintInteractionController, choosePaper paperList: [UIPrintPaper]) -> UIPrintPaper

Thanks for the pointer to use the delegate, and the code snippet! I've tested this a bit but I'm going in blind without the actual physical printer. I've logged a task to improve this and will work on it separately from this PR: https://github.com/woocommerce/woocommerce-ios/issues/11870